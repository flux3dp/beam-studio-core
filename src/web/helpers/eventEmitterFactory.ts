import EventEmitter from 'eventemitter3';

const eventEmitters = {
  'flux-id': null,
  'top-bar': null,
  'top-bar-hints': null,
  'time-estimation-button': null,
};
export default {
  createEventEmitter: (type?: string): EventEmitter => {
    if (!type) return new EventEmitter();
    if (!eventEmitters[type]) eventEmitters[type] = new EventEmitter();
    return eventEmitters[type];
  },
};
