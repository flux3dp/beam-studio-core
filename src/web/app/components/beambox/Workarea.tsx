import React from 'react';

import { ContextMenu, ContextMenuTrigger, MenuItem } from 'helpers/react-contextmenu';
import clipboard from 'app/svgedit/operations/clipboard';
import eventEmitterFactory from 'helpers/eventEmitterFactory';
import i18n from 'helpers/i18n';
import svgEditor from 'app/actions/beambox/svg-editor';
import { getSVGAsync } from 'helpers/svg-editor-helper';

let svgCanvas;
getSVGAsync((globalSVG) => { svgCanvas = globalSVG.Canvas; });

const eventEmitter = eventEmitterFactory.createEventEmitter('workarea');

interface State {
  menuDisabled: boolean,
  select: boolean,
  paste: boolean,
  group: boolean,
  ungroup: boolean,
}

export default class Workarea extends React.PureComponent<{ className: string }, State> {
  constructor(props: { className: string }) {
    super(props);
    this.state = {
      menuDisabled: false,
      select: false,
      paste: false,
      group: false,
      ungroup: false,
    };
  }

  componentDidMount(): void {
    eventEmitter.on('update-context-menu', this.updateContextMenu);
  }

  componentWillUnmount(): void {
    eventEmitter.removeListener('update-context-menu', this.updateContextMenu);
  }

  private updateContextMenu = (newValues) => {
    const newState = {
      ...this.state,
      ...newValues,
    };
    this.setState(newState);
  };

  render(): JSX.Element {
    const LANG = i18n.lang.beambox.context_menu;
    const { className } = this.props;
    const {
      menuDisabled, select, paste, group, ungroup,
    } = this.state;

    const isTouchable = navigator.maxTouchPoints >= 1;
    return (
      <>
        <ContextMenuTrigger
          id="canvas-contextmenu"
          holdToDisplay={isTouchable ? 1000 : -1}
          disable={menuDisabled}
        >
          <div id="workarea" className={className}>
            <div
              id="svgcanvas"
              style={{
                position: 'relative',
              }}
            />
          </div>
        </ContextMenuTrigger>
        <ContextMenu id="canvas-contextmenu">
          <MenuItem disabled={!select} onClick={() => svgEditor.cutSelected()}>
            {LANG.cut}
          </MenuItem>
          <MenuItem disabled={!select} onClick={() => svgEditor.copySelected()}>
            {LANG.copy}
          </MenuItem>
          <MenuItem disabled={!paste} onClick={() => clipboard.pasteElements('mouse')}>
            {LANG.paste}
          </MenuItem>
          <MenuItem disabled={!paste} onClick={() => clipboard.pasteElements('in_place')}>
            {LANG.paste_in_place}
          </MenuItem>
          <MenuItem disabled={!select} onClick={() => svgCanvas.cloneSelectedElements(20, 20)}>
            {LANG.duplicate}
          </MenuItem>
          <div className="seperator" />
          <MenuItem disabled={!select} onClick={() => svgEditor.deleteSelected()}>
            {LANG.delete}
          </MenuItem>
          <div className="seperator" />
          <MenuItem disabled={!select || !group} onClick={() => svgCanvas.groupSelectedElements()}>
            {LANG.group}
          </MenuItem>
          <MenuItem
            disabled={!select || !ungroup}
            onClick={() => svgCanvas.ungroupSelectedElement()}
          >
            {LANG.ungroup}
          </MenuItem>
          <div className="seperator" />
          <MenuItem disabled={!select} onClick={() => svgCanvas.moveTopBottomSelected('top')}>
            {LANG.move_front}
          </MenuItem>
          <MenuItem disabled={!select} onClick={() => svgCanvas.moveUpSelectedElement()}>
            {LANG.move_up}
          </MenuItem>
          <MenuItem disabled={!select} onClick={() => svgCanvas.moveDownSelectedElement()}>
            {LANG.move_down}
          </MenuItem>
          <MenuItem disabled={!select} onClick={() => svgCanvas.moveTopBottomSelected('bottom')}>
            {LANG.move_back}
          </MenuItem>
        </ContextMenu>
      </>
    );
  }
}
