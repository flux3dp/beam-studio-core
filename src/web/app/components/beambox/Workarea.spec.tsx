/* eslint-disable import/first */
import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

jest.mock('app/svgedit/operations/clipboard', () => ({
  pasteElements: jest.fn(),
}));

jest.mock('helpers/i18n', () => ({
  lang: {
    beambox: {
      context_menu: {
        cut: 'Cut',
        copy: 'Copy',
        paste: 'Paste',
        paste_in_place: 'Paste in Place',
        delete: 'Delete',
        group: 'Group',
        ungroup: 'Ungroup',
        move_front: 'Bring to Front',
        move_up: 'Bring Forward',
        move_down: 'Send Backward',
        move_back: 'Send to Back',
      },
    },
  },
}));

jest.mock('app/actions/beambox/svg-editor', () => ({
  cutSelected: jest.fn(),
  copySelected: jest.fn(),
  deleteSelected: jest.fn(),
}));

const getSVGAsync = jest.fn();
jest.mock('helpers/svg-editor-helper', () => ({
  getSVGAsync,
}));

const cloneSelectedElements = jest.fn();
const groupSelectedElements = jest.fn();
const ungroupSelectedElement = jest.fn();
const moveTopBottomSelected = jest.fn();
const moveUpSelectedElement = jest.fn();
const moveDownSelectedElement = jest.fn();

getSVGAsync.mockImplementation((callback) => {
  callback({
    Canvas: {
      cloneSelectedElements,
      groupSelectedElements,
      ungroupSelectedElement,
      moveTopBottomSelected,
      moveUpSelectedElement,
      moveDownSelectedElement,
    },
  });
});

function DummyContextMenu() {
  return (
    <div>
      This is dummy ContextMenu
    </div>
  );
}

function DummyContextMenuTrigger() {
  return (
    <div>
      This is dummy ContextMenuTrigger
    </div>
  );
}

function DummyMenuItem() {
  return (
    <div>
      This is dummy MenuItem
    </div>
  );
}

jest.mock('helpers/react-contextmenu', () => ({
  ContextMenu: DummyContextMenu,
  ContextMenuTrigger: DummyContextMenuTrigger,
  MenuItem: DummyMenuItem,
}));

import eventEmitterFactory from 'helpers/eventEmitterFactory';
import Workarea from './Workarea';

describe('test workarea', () => {
  test('should render correctly', () => {
    const eventEmitter = eventEmitterFactory.createEventEmitter('workarea');
    const wrapper = shallow(<Workarea className="mac" />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.state()).toEqual({
      menuDisabled: false,
      select: false,
      paste: false,
      group: false,
      ungroup: false,
    });
    expect(eventEmitter.eventNames().length).toBe(1);

    eventEmitter.emit('update-context-menu', { select: true, paste: true });
    expect(wrapper.state()).toEqual({
      menuDisabled: false,
      select: true,
      paste: true,
      group: false,
      ungroup: false,
    });

    eventEmitter.emit('update-context-menu', { menuDisabled: true });
    expect(wrapper.state()).toEqual({
      menuDisabled: true,
      select: true,
      paste: true,
      group: false,
      ungroup: false,
    });
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.unmount();
    expect(eventEmitter.eventNames().length).toBe(0);
  });
});
