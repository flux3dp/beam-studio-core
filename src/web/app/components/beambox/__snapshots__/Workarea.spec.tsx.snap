// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test workarea should render correctly 1`] = `
<Fragment>
  <DummyContextMenuTrigger
    disable={false}
    holdToDisplay={-1}
    id="canvas-contextmenu"
  >
    <div
      className="mac"
      id="workarea"
    >
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": "",
          }
        }
        id="styleoverrides"
        media="screen"
        scoped={true}
        type="text/css"
      />
      <div
        id="svgcanvas"
        style={
          Object {
            "position": "relative",
          }
        }
      />
    </div>
  </DummyContextMenuTrigger>
  <DummyContextMenu
    id="canvas-contextmenu"
  >
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Cut
    </DummyMenuItem>
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Copy
    </DummyMenuItem>
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Paste
    </DummyMenuItem>
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Paste in Place
    </DummyMenuItem>
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    />
    <div
      className="seperator"
    />
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Delete
    </DummyMenuItem>
    <div
      className="seperator"
    />
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Group
    </DummyMenuItem>
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Ungroup
    </DummyMenuItem>
    <div
      className="seperator"
    />
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Bring to Front
    </DummyMenuItem>
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Bring Forward
    </DummyMenuItem>
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Send Backward
    </DummyMenuItem>
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Send to Back
    </DummyMenuItem>
  </DummyContextMenu>
</Fragment>
`;

exports[`test workarea should render correctly 2`] = `
<Fragment>
  <DummyContextMenuTrigger
    disable={true}
    holdToDisplay={-1}
    id="canvas-contextmenu"
  >
    <div
      className="mac"
      id="workarea"
    >
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": "",
          }
        }
        id="styleoverrides"
        media="screen"
        scoped={true}
        type="text/css"
      />
      <div
        id="svgcanvas"
        style={
          Object {
            "position": "relative",
          }
        }
      />
    </div>
  </DummyContextMenuTrigger>
  <DummyContextMenu
    id="canvas-contextmenu"
  >
    <DummyMenuItem
      disabled={false}
      onClick={[Function]}
    >
      Cut
    </DummyMenuItem>
    <DummyMenuItem
      disabled={false}
      onClick={[Function]}
    >
      Copy
    </DummyMenuItem>
    <DummyMenuItem
      disabled={false}
      onClick={[Function]}
    >
      Paste
    </DummyMenuItem>
    <DummyMenuItem
      disabled={false}
      onClick={[Function]}
    >
      Paste in Place
    </DummyMenuItem>
    <DummyMenuItem
      disabled={false}
      onClick={[Function]}
    />
    <div
      className="seperator"
    />
    <DummyMenuItem
      disabled={false}
      onClick={[Function]}
    >
      Delete
    </DummyMenuItem>
    <div
      className="seperator"
    />
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Group
    </DummyMenuItem>
    <DummyMenuItem
      disabled={true}
      onClick={[Function]}
    >
      Ungroup
    </DummyMenuItem>
    <div
      className="seperator"
    />
    <DummyMenuItem
      disabled={false}
      onClick={[Function]}
    >
      Bring to Front
    </DummyMenuItem>
    <DummyMenuItem
      disabled={false}
      onClick={[Function]}
    >
      Bring Forward
    </DummyMenuItem>
    <DummyMenuItem
      disabled={false}
      onClick={[Function]}
    >
      Send Backward
    </DummyMenuItem>
    <DummyMenuItem
      disabled={false}
      onClick={[Function]}
    >
      Send to Back
    </DummyMenuItem>
  </DummyContextMenu>
</Fragment>
`;
